pantheon-sessions.php

```diff
@@ -342,12 +342,118 @@ public static function check_native_primary_keys() {
 
 	/**
 	 * Set id as primary key in the Native PHP Sessions plugin table.
+	 *
+	 * @param int $start_position Site index to begin from.
 	 */
-	public function add_index() {
+	public function add_index( $start_position ) {
+		global $wpdb;
+		if ( ! is_multisite() ) {
+			$this->safe_output( __( 'Single site detected. Beginning processing... \n', 'wp-native-php-sessions' ), 'log' );
+
+			$this->add_single_index( $wpdb->prefix );
+
+			$this->safe_output( __( 'Operation complete, please verify that your site is working as expected. When ready, run wp pantheon session primary-key-finalize to clean up old data, or run wp pantheon session primary-key-revert if there were issues.', 'wp-native-php-sessions' ), 'log' );
+
+			return;
+		}
+
+		$this->safe_output( __( 'Multisite installation detected. Processing Sites individually...', 'wp-native-php-sessions' ), 'log' );
+
+		$site_list = get_sites( [ 'number' => 500 ] );
+		$site_count = count( $site_list );
+
+		for ( $i = $start_position; $i < $site_count; $i++ ) {
+			// translators: %s is the current blog, and then the array index of the current site.
+			$this->safe_output( __( 'Processing blog %s. In the event of a timeout or error, resume execution starting from this point via "wp pantheon session add-index --start_point=%s". To skip this blog if it does not need processing, run "wp pantheon session add-index %s".', 'wp-native-php-sessions' ), 'log', [ $site_list[ $i ]->blog_id, $i, $i + 1 ] );
+
+			$this->add_single_index( $wpdb->get_blog_prefix( $site_list[ $i ]->blog_id ), true );
+
+			// translators: %s is the current blog id.
+			$this->safe_output( __( 'Processing for blog %s complete.', 'wp-native-php-sessions' ), 'log', [ $site_list[ $i ]->blog_id ] );
+		}
+
+		$this->safe_output( __( 'Operation complete, please verify that your site is working as expected. When ready, run wp pantheon session primary-key-finalize to clean up old data, or run wp pantheon session primary-key-revert if there were issues.', 'wp-native-php-sessions' ), 'log' );
+	}
+
+	/**
+	 * Finalizes the creation of a primary key by deleting the old data.
+	 *
+	 * @param int $start_position Site index to begin from.
+	 */
+	public function primary_key_finalize( $start_position ) {
+		global $wpdb;
+
+		if ( ! is_multisite() ) {
+			$this->safe_output( __( 'Single site detected. Beginning finalization.', 'wp-native-php-sessions' ), 'log' );
+
+			$this->add_single_index( $wpdb->prefix );
+
+			$this->safe_output( __( 'Operation complete.', 'wp-native-php-sessions' ), 'log' );
+
+			return;
+		}
+
+		$this->safe_output( __( 'Multisite installation detected. Processing Sites individually.', 'wp-native-php-sessions' ), 'log' );
+
+		$site_list = get_sites( [ 'number' => 500 ] );
```

Comment:
This and any other instances of `get_sites` still need the same treatment as above. For the record, when I use 🔢 it means this comment applies elsewhere in the file, so other places where I put that emoji mean that there are other instances of the same issue. (e.g. with the blog prefix)